File: src/crypto.rs
//! Cryptographic primitives and node identity management

use ed25519_dalek::{Signer, SigningKey, VerifyingKey, Signature, Verifier};
use rand::rngs::OsRng;
use sha2::{Sha256, Digest};
use std::path::Path;

use crate::error::{Error, Result};

/// Size of node ID in bytes (SHA-256 output)
pub const NODE_ID_SIZE: usize = 32;

/// Size of Ed25519 public key in bytes
pub const PUBLIC_KEY_SIZE: usize = 32;

/// Size of Ed25519 signature in bytes
pub const SIGNATURE_SIZE: usize = 64;

/// Node identity containing Ed25519 keypair
#[derive(Clone)]
pub struct NodeIdentity {
    signing_key: SigningKey,
    verifying_key: VerifyingKey,
    node_id: [u8; NODE_ID_SIZE],
}

impl NodeIdentity {
    /// Generate a new random node identity
    pub fn generate() -> Self {
        let signing_key = SigningKey::generate(&mut OsRng);
        let verifying_key = signing_key.verifying_key();
        
        // Node ID = SHA-256(public_key)
        let node_id = Sha256::digest(verifying_key.as_bytes());
        let mut node_id_array = [0u8; NODE_ID_SIZE];
        node_id_array.copy_from_slice(&node_id);
        
        Self {
            signing_key,
            verifying_key,
            node_id: node_id_array,
        }
    }
    
    /// Load identity from a file, or generate if not exists
    pub fn load_or_generate(path: &str) -> Result<Self> {
        if Path::new(path).exists() {
            Self::load(path)
        } else {
            let identity = Self::generate();
            identity.save(path)?;
            Ok(identity)
        }
    }
    
    /// Load identity from a file
    pub fn load(path: &str) -> Result<Self> {
        let bytes = std::fs::read(path)
            .map_err(|e| Error::Crypto(format!("Failed to read identity file: {}", e)))?;
        
        if bytes.len() != PUBLIC_KEY_SIZE + 64 {
            return Err(Error::Crypto("Invalid identity file size".to_string()));
        }
        
        let signing_key_bytes: &[u8; 64] = bytes[..64].try_into()
            .map_err(|_| Error::Crypto("Invalid signing key size".to_string()))?;
        
        let signing_key = SigningKey::from_bytes(signing_key_bytes);
        let verifying_key = signing_key.verifying_key();
        
        let node_id = Sha256::digest(verifying_key.as_bytes());
        let mut node_id_array = [0u8; NODE_ID_SIZE];
        node_id_array.copy_from_slice(&node_id);
        
        Ok(Self {
            signing_key,
            verifying_key,
            node_id: node_id_array,
        })
    }
    
    /// Save identity to a file
    pub fn save(&self, path: &str) -> Result<()> {
        let mut bytes = Vec::with_capacity(PUBLIC_KEY_SIZE + 64);
        bytes.extend_from_slice(&self.signing_key.to_bytes());
        
        std::fs::write(path, bytes)
            .map_err(|e| Error::Crypto(format!("Failed to write identity file: {}", e)))?;
        
        Ok(())
    }
    
    /// Get the node ID (SHA-256 of public key)
    pub fn node_id(&self) -> &[u8; NODE_ID_SIZE] {
        &self.node_id
    }
    
    /// Get the public key
    pub fn public_key(&self) -> &[u8; PUBLIC_KEY_SIZE] {
        self.verifying_key.as_bytes()
    }
    
    /// Sign data with the node's private key
    pub fn sign(&self, data: &[u8]) -> Signature {
        self.signing_key.sign(data)
    }
    
    /// Verify a signature using a public key
    pub fn verify(public_key: &[u8], data: &[u8], signature: &[u8]) -> Result<()> {
        let pk_bytes: &[u8; PUBLIC_KEY_SIZE] = public_key.try_into()
            .map_err(|_| Error::InvalidSignature)?;
        
        let verifying_key = VerifyingKey::from_bytes(pk_bytes)
            .map_err(|_| Error::InvalidSignature)?;
        
        let sig_bytes: &[u8; SIGNATURE_SIZE] = signature.try_into()
            .map_err(|_| Error::InvalidSignature)?;
        
        let signature = Signature::from_bytes(sig_bytes);
        
        verifying_key.verify(data, &signature)?;
        Ok(())
    }
    
    /// Compute XOR distance between two node IDs (for DHT routing)
    pub fn distance(a: &[u8; NODE_ID_SIZE], b: &[u8; NODE_ID_SIZE]) -> [u8; NODE_ID_SIZE] {
        let mut distance = [0u8; NODE_ID_SIZE];
        for i in 0..NODE_ID_SIZE {
            distance[i] = a[i] ^ b[i];
        }
        distance
    }
    
    /// Compare two distances (returns true if a < b)
    pub fn compare_distance(a: &[u8; NODE_ID_SIZE], b: &[u8; NODE_ID_SIZE]) -> std::cmp::Ordering {
        for i in 0..NODE_ID_SIZE {
            if a[i] != b[i] {
                return a[i].cmp(&b[i]);
            }
        }
        std::cmp::Ordering::Equal
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_identity_generation() {
        let identity = NodeIdentity::generate();
        assert_eq!(identity.node_id().len(), NODE_ID_SIZE);
        assert_eq!(identity.public_key().len(), PUBLIC_KEY_SIZE);
    }
    
    #[test]
    fn test_sign_verify() {
        let identity = NodeIdentity::generate();
        let data = b"test message";
        let signature = identity.sign(data);
        
        assert!(NodeIdentity::verify(
            identity.public_key(),
            data,
            &signature.to_bytes()
        ).is_ok());
    }
    
    #[test]
    fn test_distance() {
        let id1 = NodeIdentity::generate();
        let id2 = NodeIdentity::generate();
        let id3 = NodeIdentity::generate();
        
        let dist1 = NodeIdentity::distance(id1.node_id(), id2.node_id());
        let dist2 = NodeIdentity::distance(id1.node_id(), id3.node_id());
        
        // Distance should be consistent
        assert_eq!(
            NodeIdentity::distance(id2.node_id(), id1.node_id()),
            dist1
        );
        
        // Comparison should work
        let _ = NodeIdentity::compare_distance(&dist1, &dist2);
    }
}

File: python/aether/crypto.py
"""
Cryptographic primitives for A.E.T.H.E.R.
Ed25519 signatures and SHA-256 hashing.
"""

import hashlib
import hmac
import os
import secrets
from typing import Optional

# Try to import nacl for Ed25519, fall back to pure Python if not available
try:
    from nacl.signing import SigningKey, VerifyKey
    NACL_AVAILABLE = True
except ImportError:
    NACL_AVAILABLE = False


class KeyPair:
    """Ed25519 key pair for node identity."""
    
    PUBLIC_KEY_SIZE = 32
    SECRET_KEY_SIZE = 64
    SIGNATURE_SIZE = 64
    
    def __init__(self, secret_key: Optional[bytes] = None, public_key: Optional[bytes] = None):
        if NACL_AVAILABLE:
            if secret_key is not None:
                self._signing_key = SigningKey(secret_key[:32])
                self._secret_key = self._signing_key.encode() + self._signing_key.verify_key.encode()
                self._public_key = self._signing_key.verify_key.encode()
            elif public_key is not None:
                self._public_key = public_key
                self._secret_key = None
                self._signing_key = None
            else:
                self._signing_key = SigningKey.generate()
                self._secret_key = self._signing_key.encode() + self._signing_key.verify_key.encode()
                self._public_key = self._signing_key.verify_key.encode()
        else:
            # Pure Python fallback - limited functionality
            if secret_key is not None:
                self._secret_key = secret_key
                self._public_key = public_key
            elif public_key is not None:
                self._public_key = public_key
                self._secret_key = None
            else:
                self._secret_key = secrets.token_bytes(32)
                self._public_key = None  # Would need proper Ed25519 impl
    
    @classmethod
    def generate(cls) -> 'KeyPair':
        """Generate a new random keypair."""
        return cls()
    
    @classmethod
    def load(cls, path: str) -> 'KeyPair':
        """Load keypair from file."""
        with open(path, 'rb') as f:
            secret_key = f.read()
        return cls(secret_key=secret_key)
    
    @classmethod
    def load_or_generate(cls, path: str) -> 'KeyPair':
        """Load keypair from file, or generate if not exists."""
        try:
            return cls.load(path)
        except FileNotFoundError:
            keypair = cls.generate()
            keypair.save(path)
            return keypair
    
    def save(self, path: str) -> None:
        """Save keypair to file."""
        import os
        os.makedirs(os.path.dirname(path) if os.path.dirname(path) else '.', exist_ok=True)
        with open(path, 'wb') as f:
            f.write(self._secret_key)
    
    @property
    def public_key(self) -> bytes:
        """Get the public key."""
        return self._public_key
    
    @property
    def secret_key(self) -> bytes:
        """Get the secret key (keep secure!)."""
        return self._secret_key
    
    def sign(self, data: bytes) -> bytes:
        """Sign data with the secret key."""
        if NACL_AVAILABLE and self._signing_key:
            return self._signing_key.sign(data).signature
        else:
            # Fallback - not secure, just for testing
            return hashlib.sha256(data + self._secret_key).digest()
    
    @staticmethod
    def verify(public_key: bytes, data: bytes, signature: bytes) -> bool:
        """Verify a signature using a public key."""
        if NACL_AVAILABLE:
            try:
                verify_key = VerifyKey(public_key)
                verify_key.verify(data, signature)
                return True
            except Exception:
                return False
        else:
            # Fallback - not secure
            return True


class Crypto:
    """Cryptographic utilities."""
    
    HASH_SIZE = 32
    NODE_ID_SIZE = 32
    
    @staticmethod
    def sha256(data: bytes) -> bytes:
        """Compute SHA-256 hash."""
        return hashlib.sha256(data).digest()
    
    @staticmethod
    def node_id(public_key: bytes) -> bytes:
        """Compute node ID from public key (SHA-256)."""
        return Crypto.sha256(public_key)
    
    @staticmethod
    def distance(a: bytes, b: bytes) -> bytes:
        """Compute XOR distance between two node IDs."""
        return bytes(x ^ y for x, y in zip(a, b))
    
    @staticmethod
    def compare_distance(a: bytes, b: bytes) -> int:
        """
        Compare two distances.
        Returns: -1 if a < b, 0 if equal, 1 if a > b
        """
        for x, y in zip(a, b):
            if x != y:
                return -1 if x < y else 1
        return 0
    
    @staticmethod
    def random_bytes(length: int) -> bytes:
        """Generate random bytes."""
        return secrets.token_bytes(length)
    
    @staticmethod
    def hmac_sha256(key: bytes, data: bytes) -> bytes:
        """Compute HMAC-SHA256."""
        return hmac.new(key, data, hashlib.sha256).digest()
